engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobBatchCollector.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.processing.job;
+
+import io.camunda.zeebe.engine.processing.streamprocessor.TypedRecord;
+import io.camunda.zeebe.engine.state.immutable.JobState;
+import io.camunda.zeebe.engine.state.immutable.VariableState;
+import io.camunda.zeebe.msgpack.value.DocumentValue;
+import io.camunda.zeebe.msgpack.value.LongValue;
+import io.camunda.zeebe.msgpack.value.StringValue;
+import io.camunda.zeebe.msgpack.value.ValueArray;
+import io.camunda.zeebe.protocol.impl.record.value.job.JobBatchRecord;
+import io.camunda.zeebe.protocol.impl.record.value.job.JobRecord;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import java.util.Collection;
+import java.util.function.Predicate;
+import org.agrona.DirectBuffer;
+import org.agrona.ExpandableArrayBuffer;
+import org.agrona.collections.MutableInteger;
+import org.agrona.collections.MutableReference;
+import org.agrona.collections.ObjectHashSet;
+
+/**
+ * Collects jobs to be activated as part of a {@link JobBatchRecord}. Activate-able jobs are read
+ * from the {@link JobState}, resolving and setting their variables from the {@link VariableState},
+ * and added to the given batch record.
+ */
+final class JobBatchCollector {
+  private final ObjectHashSet<DirectBuffer> variableNames = new ObjectHashSet<>();
+
+  private final JobState jobState;
+  private final VariableState variableState;
+  private final Predicate<Integer> canWriteEventOfLength;
+
+  /**
+   * @param jobState the state from which jobs are collected
+   * @param variableState the state from which variables are resolved and collected
+   * @param canWriteEventOfLength a predicate which should return whether the resulting {@link
+   *     TypedRecord} containing the {@link JobBatchRecord} will be writable or not. The predicate
+   *     takes in the size of the record, and should return true if it can write such a record, and
+   *     false otherwise
+   */
+  JobBatchCollector(
+      final JobState jobState,
+      final VariableState variableState,
+      final Predicate<Integer> canWriteEventOfLength) {
```

Comment:
I went with passing a predicate instead of the full writer for two reasons. I didn't see the need to couple this class to the writer (which makes it easier to maintain/test imo), and in the end I wanted to discourage (for now anyway) writing here, and keep all the writes in the processor (as much as possible anyway). I think that makes the processing itself easier to understand in the end.